automata/fa/dfa.py

```diff
@@ -1347,6 +1364,106 @@ def from_substring(
             final_states=final_states if contains else states - final_states,
         )
 
+    @classmethod
+    def from_substrings(
+        cls: Type[Self],
+        input_symbols: AbstractSet[str],
+        substrings: AbstractSet[str],
+        *,
+        contains: bool = True,
+        must_be_suffix: bool = False,
+    ) -> Self:
+        """
+        Directly computes a DFA recognizing strings containing at least one of the given substring.
+        The implementation is based on the Aho-Corasick string-searching algorithm.
+        If contains is set to False then the complement is constructed instead.
+        If must_be_suffix is set to True, then the each substring must be a suffix instead.
+        """
+        class OutNode:
```

Comment:
Will take a closer look later, but I think some of these node classes can be simplified by using the strings themselves as state names? I believe this is constructing a trie just like in `from_finite_language`, maybe look there for ideas (assuming I'm correctly understanding how you want to simplify this implementation).

Reply:
Using the strings as state names is neither time nor memory efficient though. I would rather use integers myself. The keyword string is honestly useless in our case, it is a relic of the C++ implementation so it can be removed. At that point the OutNode is just a singly linked list which can surely be represented differently.

Reply:
You raise some good points. I just got back from traveling so I'll try to take a closer look, but in general I agree that I think there's some way the complexity of this implementation can be reduced somewhat.

Reply:
We can put it in like this and then try to reduce complexity later. It's still rather useful.
We can also add another utility function which makes use of it, constructing a DFA forbidding or requiring consecutive repetitions of identical characters.

Reply:
I think I found a bug, will try to make a small test case that produces it.

Reply:
Sounds good and no rush! Right now we're focusing on the docs changes in anticipation of the PyOpenSci submission. As that's the priority, we probably won't merge this until we get the ball rolling on that.

Reply:
There was no bug, I simply made a human error.
Implementation is fine and I have field tested it rather thoroughly.
Performance wise it has been alright so far, haven't benchmarked anything but has worked for my use case of up to 100000 strings of length in [0, 32] and a few other tests of various sizes.

Reply:
Ahh that makes sense. As long as the efficiency is reasonable, I'm ok with merging as-is and we can speed things up later if people run into issues. The main thing we need to add to this PR is a docstring, but otherwise the code is relatively clean and the test cases look good @Tagl

Reply:
What is missing from the docstring?

Reply:
If you look at the main branch, all the public functions have numpy style docstrings that describe the arguments and return value. The new documentation site is generated using these: https://github.com/caleb531/automata/blob/main/automata/fa/dfa.py#L714

Reply:
Sure, the algorithm does not need changes, but I won't be free to work on this until Friday at the earliest